File: monitor/monitor.py
import time
import json
from loguru import logger
from redis_sesion import create_sesion_redit

r = create_sesion_redit()
ACTIVE_CONTAINERS_KEY = "manager:active_containers"
WORKER_TASKS_LIST = "autoreply_queue"
HB_KEY_PREFIX = "hb:container:"
CHECK_INTERVAL = 10
MAX_SILENCE_SECONDS = 500
SLOT_IDLE_TIMEOUT = 600

# in-memory защита от ложных рестартов
container_failures = {}


def iter_hb_keys(redis):
    cursor = 0
    while True:
        cursor, keys = redis.scan(
            cursor=cursor,
            match=f"{HB_KEY_PREFIX}*",
            count=100
        )
        for k in keys:
            yield k
        if cursor == 0:
            break


def monitor_step():
    now = time.time()
    # Сет для хранения имен контейнеров, которые прислали хоть какой-то HB
    seen_containers = set()

    # --- ШАГ 1: Проверяем существующие Heartbeat ключи ---
    for key in iter_hb_keys(r):
        raw = r.get(key)
        if not raw:
            continue

        try:
            data = json.loads(raw)
            container = data["container"]
            # Добавляем в список тех, у кого есть ключ в Redis
            seen_containers.add(container)

            slots = data.get("slots", [])
            last_ts = data.get("ts", 0)

            # Проверка зависания (время в ключе устарело)
            if now - last_ts > MAX_SILENCE_SECONDS:
                cnt = container_failures.get(container, 0) + 1
                container_failures[container] = cnt
                if cnt >= 2:
                    logger.error(f"🚨 Контейнер {container} завис (старый ts). Рестарт.")
                    trigger_container_restart(container)
                    container_failures.pop(container, None)
                else:
                    logger.warning(f"⚠️ Контейнер {container} подозрительный ({cnt}/2)")
                # Пропускаем проверку слотов, если весь контейнер висит
                continue
            else:
                container_failures.pop(container, None)

            # Проверка каждого слота
            for slot in slots:
                slot_id = slot["slot_id"]
                if not slot.get("running", True):
                    trigger_slot_restart(container, slot_id)
                    continue

                last_active = slot.get("last_active", now)
                if now - last_active > SLOT_IDLE_TIMEOUT:
                    logger.error(f"⏳ Слот {slot_id} в {container} замерз")
                    trigger_slot_restart(container, slot_id)

        except Exception as e:
            logger.error(f"Ошибка парсинга HB {key}: {e}")

    # --- ШАГ 2: Проверяем контейнеры, которых ВООБЩЕ нет в Redis ---
    try:
        # Получаем список всех контейнеров, которые ДОЛЖНЫ работать
        active_names_in_redis = r.smembers(ACTIVE_CONTAINERS_KEY)
        for name in active_names_in_redis:
            name_str = name.decode() if isinstance(name, bytes) else name

            # Если контейнера нет в тех, кого мы видели на Шаге 1
            if name_str not in seen_containers:
                logger.critical(f"🚨 Контейнер {name_str} ВООБЩЕ не прислал Heartbeat! Рестарт.")
                trigger_container_restart(name_str)

    except Exception as e:
        logger.error(f"Ошибка при проверке отсутствующих HB: {e}")


def trigger_slot_restart(container, slot_id):
    """Отправляет команду на перезапуск слота в очередь менеджера"""
    # Менеджеру нужна команда 'start', чтобы запустить слот заново
    # Если слот завис, менеджер сначала должен его 'стопнуть' (очистить Redis)
    r.lpush(
        WORKER_TASKS_LIST, # Очередь, которую слушает менеджер
        json.dumps({
            "command": "restart_slot", # Новая команда для менеджера
            "slot_id": slot_id,
            "container_name": container,
            "initiated_by": "monitor"
        })
    )
    logger.info(f"🚀 Monitor: Sent restart_slot for {slot_id}")

def trigger_container_restart(container):
    """Отправляет команду на полный перезапуск контейнера"""
    r.lpush(
        WORKER_TASKS_LIST,
        json.dumps({
            "command": "restart_container", # Новая команда для менеджера
            "container_name": container,
            "initiated_by": "monitor"
        })
    )
    logger.info(f"🚀 Monitor: Sent restart_container for {container}")


if __name__ == "__main__":
    while True:
        monitor_step()
        time.sleep(CHECK_INTERVAL)

File: internal/monitor/monitor.go
// Package monitor checks container and slot health via Redis heartbeats.
package monitor

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"os/signal"
	"syscall"
	"time"

	"github.com/redis/go-redis/v9"

	"github.com/Dark-F0X/RedTailFox/internal/model"
)

const (
	heartbeatKeyPrefix   = "hb:container:"
	activeContainersKey  = "manager:active_containers"
	failureThreshold     = 2
	maxSlotRestarts      = 3
	maxContainerRestarts = 5
	defaultCheckInterval = 10 * time.Second
)

// Redis key prefixes for persisted failure counters.
const (
	failureKeyPrefix          = "monitor:failures:"
	slotFailureKeyPrefix      = "monitor:slot_failures:"
	containerRestartKeyPrefix = "monitor:restarts:"
	counterTTL                = time.Hour
)

// Config holds monitor settings.
type Config struct {
	// MaxSilenceSeconds is the maximum age of a heartbeat before considering the container stale.
	MaxSilenceSeconds int64

	// SlotIdleTimeout is the maximum inactivity duration for a slot before restarting it.
	SlotIdleTimeout int64

	// TasksQueue is the Redis list where restart commands are published.
	TasksQueue string

	// CheckInterval is the time between health check cycles.
	CheckInterval time.Duration
}

// Monitor periodically checks heartbeats and restarts unhealthy containers or slots.
//
// Failure counters are persisted in Redis so they survive monitor restarts.
// Counter keys use a TTL to prevent unbounded growth from orphaned entries.
type Monitor struct {
	rdb *redis.Client
	cfg Config
	log *slog.Logger
}

// New creates a new Monitor.
func New(rdb *redis.Client, cfg Config, log *slog.Logger) *Monitor {
	if cfg.CheckInterval == 0 {
		cfg.CheckInterval = defaultCheckInterval
	}

	return &Monitor{
		rdb: rdb,
		cfg: cfg,
		log: log,
	}
}

// Run starts the monitor loop with graceful shutdown on SIGTERM/SIGINT.
func (mon *Monitor) Run(ctx context.Context) {
	ctx, cancel := signal.NotifyContext(ctx, syscall.SIGINT, syscall.SIGTERM)
	defer cancel()

	mon.log.Info("monitor started", "interval", mon.cfg.CheckInterval)

	ticker := time.NewTicker(mon.cfg.CheckInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			mon.log.Warn("monitor shutting down")

			return
		case <-ticker.C:
			mon.Step(ctx)
		}
	}
}

// Step performs a single health check cycle. Exported for testing.
// Not safe for concurrent use — must be called from a single goroutine.
func (mon *Monitor) Step(ctx context.Context) {
	containers := mon.activeContainerNames(ctx)
	seen := mon.checkHeartbeats(ctx, containers)
	mon.checkMissingContainers(ctx, containers, seen)
}

func (mon *Monitor) checkHeartbeats(ctx context.Context, containers []string) map[string]bool {
	seen := make(map[string]bool)
	now := time.Now().Unix()

	for _, containerName := range containers {
		key := heartbeatKeyPrefix + containerName

		hbt := mon.readHeartbeat(ctx, key)
		if hbt == nil {
			continue
		}

		seen[hbt.Container] = true

		if mon.isStaleHeartbeat(now, hbt.Timestamp) {
			mon.handleStaleContainer(ctx, hbt.Container)

			continue
		}

		// Heartbeat is fresh — reset failure and restart counters.
		mon.resetCounters(ctx,
			failureKeyPrefix+hbt.Container,
			containerRestartKeyPrefix+hbt.Container,
		)

		mon.checkSlots(ctx, hbt.Container, hbt.Slots, now)
	}

	return seen
}

func (mon *Monitor) activeContainerNames(ctx context.Context) []string {
	members, err := mon.rdb.SMembers(ctx, activeContainersKey).Result()
	if err != nil {
		mon.log.Error("failed to read active containers for heartbeat check", "error", err)

		return nil
	}

	return members
}

func (mon *Monitor) readHeartbeat(ctx context.Context, key string) *model.Heartbeat {
	val, err := mon.rdb.Get(ctx, key).Result()
	if err != nil {
		mon.log.Error("failed to read heartbeat", "key", key, "error", err)

		return nil
	}

	var hbt model.Heartbeat

	err = json.Unmarshal([]byte(val), &hbt)
	if err != nil {
		mon.log.Error("failed to parse heartbeat", "key", key, "error", err)

		return nil
	}

	return &hbt
}

func (mon *Monitor) isStaleHeartbeat(now, timestamp int64) bool {
	return now-timestamp > mon.cfg.MaxSilenceSeconds
}

func (mon *Monitor) handleStaleContainer(ctx context.Context, containerName string) {
	mon.maybeRestartContainer(ctx, containerName, "stale heartbeat")
}

// maybeRestartContainer applies the failure counter + restart rate limit
// and sends a restart_container task when thresholds are met.
func (mon *Monitor) maybeRestartContainer(ctx context.Context, containerName, reason string) {
	failKey := failureKeyPrefix + containerName
	count := mon.incrCounter(ctx, failKey)

	if count < failureThreshold {
		mon.log.Warn("container health check failed",
			"container", containerName,
			"reason", reason,
			"failures", count,
			"threshold", failureThreshold,
		)

		return
	}

	restartKey := containerRestartKeyPrefix + containerName
	restarts := mon.getCounter(ctx, restartKey)

	if restarts >= maxContainerRestarts {
		mon.log.Error("container exceeded max restart attempts, manual intervention needed",
			"container", containerName,
			"reason", reason,
			"restarts", restarts,
		)

		return
	}

	mon.incrCounter(ctx, restartKey)

	mon.log.Error("container unhealthy, restarting",
		"container", containerName,
		"reason", reason,
		"failures", count,
		"restartAttempt", restarts+1,
	)

	mon.sendTask(ctx, &model.Task{
		Command:       model.CommandRestartContainer,
		ContainerName: containerName,
		InitiatedBy:   "monitor",
	})

	mon.resetCounters(ctx, failKey)
}

func (mon *Monitor) checkSlots(ctx context.Context, containerName string, slots []model.SlotHeartbeat, now int64) {
	for idx := range slots {
		slot := &slots[idx]
		key := slotFailureKey(containerName, slot.SlotID)

		if !slot.Running {
			mon.handleUnhealthySlot(ctx, key, containerName, slot.SlotID, "slot not running")

			continue
		}

		if now-slot.LastActive > mon.cfg.SlotIdleTimeout {
			mon.handleUnhealthySlot(ctx, key, containerName, slot.SlotID,
				fmt.Sprintf("slot idle %ds", now-slot.LastActive))

			continue
		}

		// Slot is healthy — reset its failure counter.
		mon.resetCounters(ctx, slotFailureKeyPrefix+key)
	}
}

func (mon *Monitor) handleUnhealthySlot(
	ctx context.Context,
	key, containerName string,
	slotID int,
	reason string,
) {
	redisKey := slotFailureKeyPrefix + key
	count := mon.incrCounter(ctx, redisKey)

	if count < failureThreshold {
		mon.log.Warn("slot health check failed",
			"container", containerName,
			"slotID", slotID,
			"reason", reason,
			"failures", count,
			"threshold", failureThreshold,
		)

		return
	}

	restartsSent := count - failureThreshold
	if restartsSent >= maxSlotRestarts {
		mon.log.Error("slot restart threshold exceeded, skipping further restarts",
			"container", containerName,
			"slotID", slotID,
			"failures", count,
			"reason", reason,
		)

		return
	}

	mon.log.Warn("unhealthy slot, sending restart",
		"container", containerName,
		"slotID", slotID,
		"failures", count,
		"reason", reason,
	)

	mon.sendSlotRestart(ctx, containerName, slotID)
}

func slotFailureKey(containerName string, slotID int) string {
	return fmt.Sprintf("%s:%d", containerName, slotID)
}

func (mon *Monitor) checkMissingContainers(ctx context.Context, containers []string, seen map[string]bool) {
	for _, name := range containers {
		if seen[name] {
			continue
		}

		mon.maybeRestartContainer(ctx, name, "missing heartbeat")
	}
}

func (mon *Monitor) sendSlotRestart(ctx context.Context, containerName string, slotID int) {
	mon.sendTask(ctx, &model.Task{
		Command:       model.CommandRestartSlot,
		SlotID:        slotID,
		ContainerName: containerName,
		InitiatedBy:   "monitor",
	})
}

func (mon *Monitor) sendTask(ctx context.Context, task *model.Task) {
	data, err := json.Marshal(task)
	if err != nil {
		mon.log.Error("failed to marshal task", "error", err)

		return
	}

	err = mon.rdb.LPush(ctx, mon.cfg.TasksQueue, data).Err()
	if err != nil {
		mon.log.Error("failed to send task", "error", err)
	}
}

// incrCounter atomically increments a Redis counter and refreshes its TTL.
// Returns the new value, or 0 on error.
func (mon *Monitor) incrCounter(ctx context.Context, key string) int {
	val, err := mon.rdb.Incr(ctx, key).Result()
	if err != nil {
		mon.log.Error("failed to increment counter", "key", key, "error", err)

		return 0
	}

	mon.rdb.Expire(ctx, key, counterTTL)

	return int(val)
}

// getCounter reads a Redis counter value. Returns 0 on error or missing key.
func (mon *Monitor) getCounter(ctx context.Context, key string) int {
	val, err := mon.rdb.Get(ctx, key).Int()
	if err != nil {
		return 0
	}

	return val
}

// resetCounters deletes the given Redis counter keys.
func (mon *Monitor) resetCounters(ctx context.Context, keys ...string) {
	if len(keys) == 0 {
		return
	}

	err := mon.rdb.Del(ctx, keys...).Err()
	if err != nil {
		mon.log.Error("failed to reset counters", "keys", keys, "error", err)
	}
}
